File: Python/spiral-matrix.py
class Solution:
    # @param {integer[][]} matrix
    # @return {integer[]}
    def spiralOrder(self, matrix):
        if not matrix: return []
        r = len(matrix)
        c = len(matrix[0])
        left, right, top, down = 0, c-1, 0, r-1
        ans = []
        
        while left<right and top<down:
            for i in xrange(right-left): ans.append(matrix[top][left+i])
            for i in xrange(down-top): ans.append(matrix[top+i][right])
            for i in xrange(right-left): ans.append(matrix[down][right-i])
            for i in xrange(down-top): ans.append(matrix[down-i][top])
            left += 1; right -= 1; top += 1; down -= 1

        if top==down:
            for i in xrange(right-left+1): ans.append(matrix[top][left+i])
        elif right==left:
            for i in xrange(down-top+1): ans.append(matrix[top+i][left])
        
        return ans

File: C++/spiral-matrix.cpp
class Solution {
public:
    std::vector<int> spiralOrder(std::vector<std::vector<int>>& matrix) {
        std::vector<int> ans;
        if(matrix.size()==0) return ans;
        int r = matrix.size();
        int c = matrix[0].size();
        int left=0, right=c-1, top=0, down=r-1;
        while(left<right && top<down) {
            for(int i=0; i<right-left; ++i) ans.push_back(matrix[top][left+i]);
            for(int i=0; i<down-top; ++i) ans.push_back(matrix[top+i][right]);
            for(int i=0; i<right-left; ++i) ans.push_back(matrix[down][right-i]);
            for(int i=0; i<down-top; ++i) ans.push_back(matrix[down-i][left]);
            ++left; --right; ++top; --down;
        }
        if(top==down) {
            for(int i=0; i<right-left+1; ++i) ans.push_back(matrix[top][left+i]);
        }
        else if(left==right) {
            for(int i=0; i<down-top+1; ++i) ans.push_back(matrix[top+i][left]);
        }
        return ans;
    }
};

